app/post_upgrade.go

```diff
@@ -50,34 +60,62 @@ func PostUpgradeCmd() cli.Command {
 
 func postUpgrade(c *cli.Context) error {
 	namespace := c.String(FlagNamespace)
-	if namespace == "" {
-		return errors.New("namespace is required")
-	}
 
 	config, err := clientcmd.BuildConfigFromFlags("", c.String(FlagKubeConfig))
 	if err != nil {
 		return errors.Wrap(err, "failed to get client config")
 	}
 
+	eventBroadcaster, err := createEventBroadcaster(config)
+	if err != nil {
+		return errors.Wrap(err, "failed to create event broadcaster")
+	}
+
+	scheme := runtime.NewScheme()
+	if err := longhorn.SchemeBuilder.AddToScheme(scheme); err != nil {
+		return errors.Wrap(err, "failed to create scheme")
+	}
+
+	eventRecorder := eventBroadcaster.NewRecorder(scheme, corev1.EventSource{Component: PostUpgradeEventer})
+
 	kubeClient, err := kubernetes.NewForConfig(config)
 	if err != nil {
 		return errors.Wrap(err, "failed to get k8s client")
 	}
 
-	return newPostUpgrader(namespace, kubeClient).Run()
+	err = newPostUpgrader(namespace, kubeClient, eventRecorder).Run()
+	if err != nil {
+		// hang around so logs cam be collected.
+		// TODO - make this a --ttl argument.
+		time.Sleep(1 * time.Hour)
+	}
```

Comment:
Note that with the event captured, this is not as necessary, but it might be useful.  I'm not sure what to use for the time to wait.  The pre-upgrade job itself has `spec.activeDeadlineSeconds: 900` so the pod will be killed after 15 minutes anyway, and perhaps that is a reasonable value to use.

Reply:
If the event is emitted, does it need to sleep for minutes?

Reply:
Not really.  The event will last for an hour, so if a support bundle is collected in that time, the event should be there.  
It would be the way to accomplish the goal in https://github.com/longhorn/longhorn/issues/9448.
I can certainly take it out if preferred.

Reply:
I removed it.

Reply:
I put it back.  Without it, sometimes the panic from the "fatal" error means the event does not get created.

Reply:
> I put it back. Without it, sometimes the panic from the "fatal" error means the event does not get created.

Can you elaborate more on the statement?

Reply:
> I put it back. Without it, sometimes the panic from the "fatal" error means the event does not get created.

Looks like the AI suggestion can solve this one. WDYT @james-munson ?

https://github.com/longhorn/longhorn-manager/pull/3211/files#r1803865276

Reply:
The events are queued by Event/Eventf, but not necessarily propagated to the sink before the Event() call returns.  They can be lost when the next thing that happens is an os.Exit as part of the log.Fatal.
However, the AI suggestion is a good one.  I tested it, and it looks like eventBroadcaster.Shutdown() forces a flush of the queued events, so they show up even without a sleep to delay the exit.  I have pushed up the change.